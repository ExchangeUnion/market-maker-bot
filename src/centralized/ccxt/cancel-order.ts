import { Exchange, Order } from 'ccxt';
import { from, Observable, defer } from 'rxjs';
import { Config } from '../../config';

const cancelOrder$ = (
  exchange: Exchange,
  config: Config,
  orderId: string
): Observable<Order> => {
  return defer(() =>
    from(
      exchange.cancelOrder(
        orderId,
        `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`
      )
    )
  );
};

export { cancelOrder$ };
