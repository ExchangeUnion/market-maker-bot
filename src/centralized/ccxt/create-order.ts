import { Exchange, Order } from 'ccxt';
import { OrderSide } from '../../constants';
import BigNumber from 'bignumber.js';
import { Observable, from, defer } from 'rxjs';
import { Config } from '../../config';

type CreateOrderParams = {
  config: Config;
  exchange: Exchange;
  side: OrderSide;
  quantity: BigNumber;
};

const createOrder$ = ({
  config,
  exchange,
  side,
  quantity,
}: CreateOrderParams): Observable<Order> => {
  return defer(() => {
    const price = undefined;
    const params =
      config.CEX === 'Kraken' ? { trading_agreement: 'agree' } : undefined;
    return from(
      exchange.createMarketOrder(
        `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`,
        side,
        parseFloat(quantity.toFixed(4, BigNumber.ROUND_DOWN)),
        price,
        params
      )
    );
  });
};

export { createOrder$, CreateOrderParams };
