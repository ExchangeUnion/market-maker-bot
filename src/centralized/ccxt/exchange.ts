import ccxt, { Exchange } from 'ccxt';
import { Config } from '../../config';

const getExchange = (config: Config): Exchange => {
  switch (config.CEX) {
    case 'BINANCE':
      return new ccxt.binance({
        apiKey: config.CEX_API_KEY,
        secret: config.CEX_API_SECRET,
      });
    case 'KRAKEN':
      return new ccxt.kraken({
        apiKey: config.CEX_API_KEY,
        secret: config.CEX_API_SECRET,
      });
    default:
      throw new Error(
        'Could not get centralized exchange. Invalide CEX configuration option'
      );
  }
};

export { getExchange };
