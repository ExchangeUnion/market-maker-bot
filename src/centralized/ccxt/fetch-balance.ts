import { Balances, Exchange } from 'ccxt';
import { defer, from, Observable } from 'rxjs';

const fetchBalance$ = (exchange: Exchange): Observable<Balances> => {
  return defer(() => from(exchange.fetchBalance()));
};

export { fetchBalance$ };
