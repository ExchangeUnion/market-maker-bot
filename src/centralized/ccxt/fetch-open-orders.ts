import { Exchange, Order } from 'ccxt';
import { defer, from, Observable } from 'rxjs';
import { Config } from '../../config';

const fetchOpenOrders$ = (
  exchange: Exchange,
  config: Config
): Observable<Order[]> => {
  return defer(() =>
    from(
      exchange.fetchOpenOrders(
        `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`
      )
    )
  );
};

export { fetchOpenOrders$ };
