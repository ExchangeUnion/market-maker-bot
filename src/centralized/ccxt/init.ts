import { Dictionary, Exchange, Market } from 'ccxt';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Config } from '../../config';

type InitCEXparams = {
  config: Config;
  loadMarkets$: (exchange: Exchange) => Observable<Dictionary<Market>>;
  getExchange: (config: Config) => Exchange;
};

type InitCEXResponse = {
  markets: Dictionary<Market>;
  exchange: Exchange;
};

const initCEX$ = ({
  getExchange,
  config,
  loadMarkets$,
}: InitCEXparams): Observable<InitCEXResponse> => {
  const exchange = getExchange(config);
  return loadMarkets$(exchange).pipe(
    map(markets => {
      return {
        markets,
        exchange,
      };
    })
  );
};

export { initCEX$, InitCEXparams, InitCEXResponse };
