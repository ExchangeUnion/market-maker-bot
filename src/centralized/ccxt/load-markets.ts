import { Dictionary, Exchange, Market } from 'ccxt';
import { defer, from, Observable } from 'rxjs';

const loadMarkets$ = (exchange: Exchange): Observable<Dictionary<Market>> => {
  return defer(() => from(exchange.loadMarkets()));
};

export { loadMarkets$ };
