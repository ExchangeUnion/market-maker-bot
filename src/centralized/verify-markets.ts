import { Dictionary, Market } from 'ccxt';
import { Config } from '../config';
import { errors } from '../opendex/errors';

const verifyMarkets = (config: Config, CEXmarkets: Dictionary<Market>) => {
  const tradingPair = `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`;
  const market = CEXmarkets[tradingPair];
  if (market && market.active) {
    return true;
  }
  throw errors.CEX_INVALID_TRADING_PAIR(tradingPair, config.CEX);
};

export { verifyMarkets };
