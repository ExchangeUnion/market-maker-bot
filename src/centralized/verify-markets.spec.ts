import { testConfig } from '../test-utils';
import { verifyMarkets } from './verify-markets';
import { Dictionary, Market } from 'ccxt';

describe('verifyMarkets', () => {
  const CEXmarkets: Dictionary<Market> = {
    'BTC/USD': {
      active: true,
    } as Market,
  };

  it('throws when BTC/USD trading pair not exist', () => {
    expect.assertions(1);
    const markets = {
      ...CEXmarkets,
    };
    delete markets['BTC/USD'];
    const config = {
      ...testConfig(),
      ...{ CEX_BASEASSET: 'BTC', CEX_QUOTEASSET: 'USD' },
    };
    expect(() => {
      verifyMarkets(config, markets);
    }).toThrowErrorMatchingSnapshot();
  });

  it.only('throws when BTC/USD trading pair is inactive', () => {
    expect.assertions(1);
    const markets = {
      ...CEXmarkets,
    };
    delete markets['BTC/USD'].active;
    const config = {
      ...testConfig(),
      ...{ CEX_BASEASSET: 'BTC', CEX_QUOTEASSET: 'USD' },
    };
    expect(() => {
      verifyMarkets(config, markets);
    }).toThrowErrorMatchingSnapshot();
  });

  it('returns true when BTC/USD trading pair exist', () => {
    expect.assertions(1);
    const config = {
      ...testConfig(),
      ...{ CEX_BASEASSET: 'BTC', CEX_QUOTEASSET: 'USD' },
    };
    expect(verifyMarkets(config, CEXmarkets)).toEqual(true);
  });
});
