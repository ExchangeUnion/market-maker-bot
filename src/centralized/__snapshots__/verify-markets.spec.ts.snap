// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifyMarkets throws when BTC/USD trading pair is inactive 1`] = `"The configured trading pair BTC/USD does not exist on Binance"`;

exports[`verifyMarkets throws when BTC/USD trading pair not exist 1`] = `"The configured trading pair BTC/USD does not exist on Binance"`;
